[SDXL DreamBooth LoRA] add support for text encoder fine-tuning

sayakpaul opened the pull request:
This PR adds support for text encoder fine-tuning in the DreamBooth LoRA script for SDXL. 

Summary of the changes:

* Major refactor of the dataloader and the collator to accommodate for training the text encoders.
* Support for the numerically stable VAE (some type-casting here and there) (#4083). 
* Changes to the loaders of LoRA. 

To help us maintain sanity, I tested the current training script under three settings:

1. No text encoder and no better VAE

```bash
export MODEL_NAME="diffusers/stable-diffusion-xl-base-0.9"
export INSTANCE_DIR="dog"
export OUTPUT_DIR="lora-trained-xl-no-vae-text-encoder"

accelerate launch train_dreambooth_lora_sdxl.py \
  --pretrained_model_name_or_path=$MODEL_NAME  \
  --instance_data_dir=$INSTANCE_DIR \
  --output_dir=$OUTPUT_DIR \
  --mixed_precision="fp16" \
  --instance_prompt="a photo of sks dog" \
  --resolution=1024 \
  --train_batch_size=1 \
  --gradient_accumulation_steps=4 \
  --learning_rate=1e-4 \
  --report_to="wandb" \
  --lr_scheduler="constant" \
  --lr_warmup_steps=0 \
  --max_train_steps=75 \
  --validation_prompt="A photo of sks dog in a bucket" \
  --validation_epochs=25 \
  --seed="0" \
  --push_to_hub
```

Artifacts:

* WandB [logs](https://wandb.ai/sayakpaul/dreambooth-lora-sd-xl/runs/6wjon1b5)
* [Checkpoints](https://huggingface.co/fusing/lora-trained-xl-no-text-encoder-no-vae) (private to the Diffusers team)

2. No text encoder but better VAE

```bash
export MODEL_NAME="diffusers/stable-diffusion-xl-base-0.9"
export INSTANCE_DIR="dog"
export OUTPUT_DIR="lora-trained-xl-no-text-encoder"

accelerate launch train_dreambooth_lora_sdxl.py \
  --pretrained_model_name_or_path=$MODEL_NAME  \
  --pretrained_vae_model_name_or_path="sayakpaul/sdxl-vae-fp16-fix-testing" \
  --instance_data_dir=$INSTANCE_DIR \
  --output_dir=$OUTPUT_DIR \
  --mixed_precision="fp16" \
  --instance_prompt="a photo of sks dog" \
  --resolution=1024 \
  --train_batch_size=1 \
  --gradient_accumulation_steps=4 \
  --learning_rate=1e-4 \
  --report_to="wandb" \
  --lr_scheduler="constant" \
  --lr_warmup_steps=0 \
  --max_train_steps=75 \
  --train_text_encoder \
  --validation_prompt="A photo of sks dog in a bucket" \
  --validation_epochs=25 \
  --seed="0" \
  --push_to_hub
```

Artifacts:

* WandB [logs](https://wandb.ai/sayakpaul/dreambooth-lora-sd-xl/runs/3yvleah0)
* [Checkpoints](https://huggingface.co/fusing/lora-trained-xl-no-text-encoder) (private to the Diffusers team)

3. Better VAE along with the text encoder

```bash
export MODEL_NAME="diffusers/stable-diffusion-xl-base-0.9"
export INSTANCE_DIR="dog"
export OUTPUT_DIR="lora-trained-xl-text-encoder-vae"

accelerate launch train_dreambooth_lora_sdxl.py \
  --pretrained_model_name_or_path=$MODEL_NAME  \
  --pretrained_vae_model_name_or_path="sayakpaul/sdxl-vae-fp16-fix-testing" \
  --instance_data_dir=$INSTANCE_DIR \
  --output_dir=$OUTPUT_DIR \
  --mixed_precision="fp16" \
  --instance_prompt="a photo of sks dog" \
  --resolution=1024 \
  --train_batch_size=1 \
  --gradient_accumulation_steps=4 \
  --learning_rate=1e-4 \
  --report_to="wandb" \
  --lr_scheduler="constant" \
  --lr_warmup_steps=0 \
  --max_train_steps=75 \
  --train_text_encoder \
  --validation_prompt="A photo of sks dog in a bucket" \
  --validation_epochs=25 \
  --seed="0" \
  --push_to_hub
```

Artifacts:

* WandB [logs](https://wandb.ai/sayakpaul/dreambooth-lora-sd-xl/runs/009mzi1a)
* [Checkpoints](https://huggingface.co/fusing/lora-trained-xl-text-enc) (private to the Diffusers team)

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/diffusers/pr_4097). All of your documentation changes will be reflected on that endpoint.

BitPhinix commented:
Is there any reason for all the trained Loras to be this "grainy"? The trained Lora (even with the dog example) makes the model "feel" worse overall.

sayakpaul commented:
> Is there any reason for all the trained Loras to be this "grainy"? The trained Lora (even with the dog example) makes the model "feel" worse overall.

They were all trained for just 75 steps and with barebones hyperparameters. So, please take those results with a grain of salt.

BitPhinix commented:
Hmm, I don't know if that is the case. Comparing two loras trained on the same settings between the kohya and diffusers implementation, the Kohya one doesn't show these artifacts. Will do a bit more digging but I suspect something is off here.

patrickvonplaten commented:
@BitPhinix, it's always helpful to provide code examples / snippets, did you find any differences in the setup? ComfyUI Dreambooth is based on `diffusers` under the hood, so it's probably quite easy to find the differences here

patrickvonplaten commented on examples/dreambooth/train_dreambooth_lora_sdxl.py:
```diff
@@ -1132,8 +1240,22 @@ def compute_embeddings(prompt, text_encoders, tokenizers):
                 pipeline = DiffusionPipeline.from_pretrained(
                     args.pretrained_model_name_or_path,
+                    vae=vae,
+                    text_encoder=accelerator.unwrap_model(text_encoder_one)
+                    if args.train_text_encoder
+                    else text_encoder_one,
+                    text_encoder_2=accelerator.unwrap_model(text_encoder_two)
+                    if args.train_text_encoder
+                    else text_encoder_two,
```
```suggestion
                    text_encoder=accelerator.unwrap_model(text_encoder_one),
                    text_encoder_2=accelerator.unwrap_model(text_encoder_two),
```

sayakpaul replied:
This is not what we want when not training the text encoder actually. 

When not training the text encoder, we precompute the text embeddings and delete the text encoder from memory to save memory. 

This is why during the final inference, we load the text encoders again (only when text encoder is not being traind). 

Makes sense?

patrickvonplaten replied:
It's just to make the code more compact - if text_encoder_one is not a `nn.DataParallel` `accelerator.unwrap_model(...)` is simply a no-op. 

So this suggestion is really just a nit to make the code more readable

patrickvonplaten commented on examples/dreambooth/train_dreambooth_lora_sdxl.py:
```diff
@@ -1189,16 +1313,34 @@ def compute_embeddings(prompt, text_encoders, tokenizers):
         unet = unet.to(torch.float32)
```
Why do we have a `.to(torch.float32)` here actually? I don't think we need this no? 
```suggestion
```

sayakpaul replied:
We always serialize in full precision unless specified otherwise.

patrickvonplaten commented on examples/dreambooth/train_dreambooth_lora_sdxl.py:
```diff
@@ -1189,16 +1313,34 @@ def compute_embeddings(prompt, text_encoders, tokenizers):
         unet_lora_layers = unet_attn_processors_state_dict(unet)
 
+        if args.train_text_encoder:
+            text_encoder_one = accelerator.unwrap_model(text_encoder_one)
+            text_encoder_one = text_encoder_one.to(torch.float32)
+            text_encoder_two = accelerator.unwrap_model(text_encoder_two)
+            text_encoder_two = text_encoder_two.to(torch.float32)
```
```suggestion
            text_encoder_one = accelerator.unwrap_model(text_encoder_one)
            text_encoder_two = accelerator.unwrap_model(text_encoder_two)
```

sayakpaul replied:
https://github.com/huggingface/diffusers/pull/4097/files#r1265134516

patrickvonplaten commented on src/diffusers/loaders.py:
```diff
@@ -830,6 +830,11 @@ def load_lora_weights(self, pretrained_model_name_or_path_or_dict: Union[str, Di
+            )
 
     @classmethod
     def lora_state_dict(
```
Can we add a test here showing that a two text encoder LoRA can be loaded?

sayakpaul replied:
Works for you? 

https://github.com/huggingface/diffusers/pull/4097/files#diff-db771746ed12c66b4bcf74a5fe059cfb9c170902d60bbce65886992a9bee34d4R388

patrickvonplaten replied:
I thought more about just a quick fast test that loads and unloads a dummy LoRA that is in the format of SD-XL

sayakpaul replied:
Added in `test_lora_layers.py`.

patrickvonplaten reviewed (commented):
Looks very nice!

williamberman commented on examples/dreambooth/train_dreambooth_lora_sdxl.py:
```diff
@@ -453,6 +455,10 @@ def __init__(
         self.instance_unet_added_conditions = instance_unet_added_conditions
         self.class_unet_added_conditions = class_unet_added_conditions
 
+        if tokenizers is not None:
+            self.tokenizer_one = tokenizers[0]
+            self.tokenizer_two = tokenizers[1]
+
```
if we know the tokenizers are going to be passed as a list of two elements, we should just pass both tokenizers as separate arguments

williamberman commented on examples/dreambooth/train_dreambooth_lora_sdxl.py:
```diff
@@ -496,7 +502,11 @@ def __getitem__(self, index):
+        if self.instance_prompt_hidden_states is not None:
+            example["instance_prompt_ids"] = self.instance_prompt_hidden_states
+        else:
+            example["instance_prompt_tokens_one"] = tokenize_prompt(self.tokenizer_one, self.instance_prompt)
+            example["instance_prompt_tokens_two"] = tokenize_prompt(self.tokenizer_two, self.instance_prompt)
```
This will re-tokenize the same prompt every time we pull from the dataloader. I think we can just tokenize it ahead of time.

williamberman commented on examples/dreambooth/train_dreambooth_lora_sdxl.py:
```diff
@@ -506,45 +516,69 @@ def __getitem__(self, index):
-        "pixel_values": pixel_values,
-        "unet_added_conditions": {"text_embeds": add_text_embeds, "time_ids": add_time_ids},
-    }
+    if not train_text_encoder:
+        input_ids = torch.cat(input_ids, dim=0)
+        add_text_embeds = torch.cat(add_text_embeds, dim=0)
+    else:
+        tokens_one = torch.cat(tokens_one, dim=0)
+        tokens_two = torch.cat(tokens_two, dim=0)
+
+    unet_added_conditions = {"time_ids": add_time_ids}
+    if not train_text_encoder:
+        unet_added_conditions.update({"text_embeds": add_text_embeds})
+        batch = {"input_ids": input_ids, "pixel_values": pixel_values, "unet_added_conditions": unet_added_conditions}
+    else:
+        batch = {
+            "tokens_one": tokens_one,
+            "tokens_two": tokens_two,
+            "pixel_values": pixel_values,
+            "unet_added_conditions": unet_added_conditions,
+        }
```
I believe there's actually a different refactor that could make this script (and a few of the other ones) a bit simpler.

The only data that belongs in the dataloader is the image data. All other outputs of the collate function, `input_ids`, `unet_added_conditions`, `attention_mask`, `tokens_one`, and `tokens_two` are statically determined and don't require any loading from files, etc...

If you were to just remove those components from the dataloader, we could simplify the dataloader and the PR could instead directly handle the text prompt components in the script body, thoughts?

sayakpaul replied:
I like the idea, however to have self-containment in the script, I think the dataloaders are okay as is.

williamberman replied:
Sorry but I'm not really sure I follow, could you elaborate on what this has to do with the script being self-contained? I'm not recommending making changes to the dataloaders in other scripts, just noting that the other scripts share the same issue.

I feel relatively strongly that we should not be passing already computed data to the dataloader

williamberman commented on src/diffusers/loaders.py:
```diff
@@ -1027,14 +1032,21 @@ def load_lora_into_text_encoder(cls, state_dict, network_alpha, text_encoder, lo
+        if any(cls.text_encoder_name in key for key in keys):
             # Load the layers corresponding to text encoder and make necessary adjustments.
-            text_encoder_keys = [k for k in keys if k.startswith(cls.text_encoder_name)]
+            if isinstance(text_encoder, CLIPTextModel):
+                prefix = cls.text_encoder_name
+            elif isinstance(text_encoder, CLIPTextModelWithProjection):
+                prefix = f"{cls.text_encoder_name}_2"
+
```
Hmm, I think this check is a bit brittle and will break other pipelines which happen to have a CLIPTextModelWithProjection class for their `text_encoder` key. I don't know if we actively support lora training for any of them (karlo is the only one I can remember off the top of my head), but regardless this is a very implicit contract that is easy to break on accident. 

It would be best if the function instead took the prefix in the state dict as an argument defaulting to `text_encoder`. Then the sdxl pipeline can override `load_lora_weights`. This also gets rid of the additional `is_sdxl` check in the `load_lora_weights` function. 


i.e.

```
class LoraLoaderMixin:
    def load_lora_weights():
        # leave as is

    def load_lora_into_text_encoder(..., prefix='text_encoder'):
        # replace `cls.text_encoder_name` in body with `prefix`
        
class SDXLPipeline:
    def load_lora_weights():
       # load unet ...
        self.load_lora_into_text_encoder(self.text_encoder, 'text_encoder')
        self.load_lora_into_text_encoder(self.text_encoder_2, 'text_encoder')
```

A good rule of thumb is any time we see ourselves looking at the structure of keys or weights in the state dict to implicitly figure out something about the model, ask if we can instead explicitly handle the case in the code and leave the contents of the state dict opaque.

williamberman commented on src/diffusers/loaders.py:
```diff
@@ -1200,16 +1217,25 @@ def save_function(weights, filename):
             state_dict.update(unet_lora_state_dict)
 
         if text_encoder_lora_layers is not None:
-            weights = (
-                text_encoder_lora_layers.state_dict()
-                if isinstance(text_encoder_lora_layers, torch.nn.Module)
-                else text_encoder_lora_layers
-            )
 
-            text_encoder_lora_state_dict = {
-                f"{self.text_encoder_name}.{module_name}": param for module_name, param in weights.items()
-            }
-            state_dict.update(text_encoder_lora_state_dict)
+            def get_te_params(te_state_dict):
+                weights = te_state_dict.state_dict() if isinstance(te_state_dict, torch.nn.Module) else te_state_dict
+                return weights
+
+            if not isinstance(text_encoder_lora_layers, list):
+                weights = get_te_params(text_encoder_lora_layers)
+                text_encoder_lora_state_dict = {
+                    f"{self.text_encoder_name}.{module_name}": param for module_name, param in weights.items()
+                }
+                state_dict.update(text_encoder_lora_state_dict)
+            else:
+                for i in range(len(text_encoder_lora_layers)):
+                    weights = get_te_params(text_encoder_lora_layers[i])
+                    weight_suffix = f"{self.text_encoder_name}_{i+1}" if i > 0 else self.text_encoder_name
+                    text_encoder_lora_state_dict = {
+                        f"{weight_suffix}.{module_name}": param for module_name, param in weights.items()
+                    }
+                    state_dict.update(text_encoder_lora_state_dict)
```
I think this is quite implicit and relies on a naming convention in the sub class, I think it would be better if we just overrode the method in the sdxl pipeline and moved the file writing code to a separate classmethod

i.e.

```py
class LoraLoaderMixin:
    @classmethod
    def save_lora_weights(...):
        state_dict = ...
        self.write_lora_weights(state_dict, ...)

   @classmethod
   def write_lora_weights(state_dict):
       ...
        
class SDXLPipeline:
    @classmethod
    def save_lora_weights(..., unet_lora_layers, text_encoder_lora_layers, text_encoder_2_lora_layers):
         state_dict = {}
         state_dict.update({f"unet.{k}": v for k, v in unet_lora_layers.state_dict()})
         state_dict.update({f"text_encoder.{k}": v for k, v in text_encoder_lora_layers.state_dict()})
         state_dict.update({f"text_encoder_2.{k}": v for k, v in text_encoder_lora_layers.state_dict()})

         self.write_lora_weights(state_dict, ...)
```

A good rule of thumb is that we should ideally treat the state dict as opaque as possible -- i.e. we just write it and load from it. Any time we have to manually either look inside it or choose what keys to write to it, be as explicit as possible.

sayakpaul replied:
Ah, I think we can still keep the changes if we introduced a `prefix` argument in `save_lora_weights()` function following https://github.com/huggingface/diffusers/pull/4097/files#r1266022763? I think that still respects what you're trying to achieve here. 

@patrickvonplaten, would love to hear your thoughts too.

patrickvonplaten replied:
Hmm yeah this one is a bit tricky. Tend to agree with @williamberman here that creating a new function seems to be most sensible here given that SD-XL is arguably a big exception since it has two text encoders.

sayakpaul replied:
If we go via this way, then we will need to initialize the SDXL pipeline in the training script just for serialization purposes. This also applies to the hooks we use for checkpointing. Are we okay with that?

williamberman replied:
I don't see what requires instantiating an instance of the class, they're all classmethods. I'm also not really sure I see how this impacts tests unfortunately. 

@sayakpaul could you show the pseudocode for how passing a prefix fixes it?

williamberman replied:
Ok I see the existing update that adds the prefix. I'm not sure I really see the difference between as it was before with adding the `text_encoder_prefix` argument, the body of the function is almost the same just the value is in a function argument? We still have the same implicit contract that the sub class must name its text encoders a particular way.

I feel relatively strongly that we should do as recommended here https://github.com/huggingface/diffusers/pull/4097#discussion_r1266035100

williamberman commented:
Nice start! I think with a few design tweaks we're good to go :)

williamberman commented on src/diffusers/loaders.py:
```diff
@@ -1087,9 +1089,12 @@ def load_lora_into_text_encoder(cls, state_dict, network_alpha, text_encoder, lo
                 load_state_dict_results = text_encoder.load_state_dict(text_encoder_lora_state_dict, strict=False)
                 if len(load_state_dict_results.unexpected_keys) != 0:
-                    raise ValueError(
-                        f"failed to load text encoder state dict, unexpected keys: {load_state_dict_results.unexpected_keys}"
-                    )
+                    unexpected_keys = load_state_dict_results.unexpected_keys
+                    is_sdxl_ckpt = any("text_encoder_" in k for k in unexpected_keys)
+                    if not is_sdxl_ckpt:
+                        raise ValueError(
+                            f"failed to load text encoder state dict, unexpected keys: {load_state_dict_results.unexpected_keys}"
+                        )
```
We can avoid this adhoc check for which model we're loading by passing in a dict with only the keys of the expected text_encoder :)

i.e.

```
class SDXLPipelin:
    def load_lora_weights():
        state_dict, network_alpha = self.lora_state_dict(pretrained_model_name_or_path_or_dict, **kwargs)
        self.load_lora_into_unet(state_dict, network_alpha=network_alpha, unet=self.unet)
        
        text_encoder_state_dict = { k: v for k, v in state_dict if 'text_encoder.' in k }
        self.load_lora_into_text_encoder(
            text_encoder_state_dict,
            network_alpha=network_alpha,
            text_encoder=self.text_encoder,
            prefix="text_encoder",
            lora_scale=self.lora_scale,
        )

       text_encoder_2_state_dict = { k: v for k, v in state_dict if 'text_encoder_2.' in k }
        self.load_lora_into_text_encoder(
            text_encoder_2_state_dict,
            network_alpha=network_alpha,
            text_encoder=self.text_encoder_2,
            prefix="text_encoder_2",
            lora_scale=self.lora_scale,
        )
```

We really want to as much as possible avoid checking what's in the state dict. If we can do that checking in the code specific to that particular model, we should :)

williamberman commented on src/diffusers/loaders.py:
```diff
@@ -1099,6 +1104,8 @@ def lora_scale(self) -> float:
 
     def _remove_text_encoder_monkey_patch(self):
         self._remove_text_encoder_monkey_patch_classmethod(self.text_encoder)
+        if hasattr(self, "text_encoder_2"):
+            self._remove_text_encoder_monkey_patch_classmethod(self.text_encoder_2)
```
let's move the sdxl pipeline specific code to the sdxl pipeline

i.e.

```py
class SDXLPipeline:
    def _remove_text_encoder_monkey_patch(self):
         self._remove_text_encoder_monkey_patch_classmethod(self.text_encoder)
         self._remove_text_encoder_monkey_patch_classmethod(self.text_encoder_2)
```

and then we don't have to touch the _remove_text_encoder_monkey_patch method in the general mixin

sayakpaul commented:
@patrickvonplaten @williamberman I think I have addressed all your comments:

* Simplification of the dataloader
* Less state dict munging 

I would suggest taking another deeper look.

patrickvonplaten commented on src/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl.py:
```diff
@@ -809,3 +810,66 @@ def __call__(
 
         return StableDiffusionXLPipelineOutput(images=image)
+
+    # Overrride to properly handle the loading and unloading of the additional text encoder.
```
ok for me!

patrickvonplaten commented on examples/dreambooth/train_dreambooth_lora_sdxl.py:
```diff
@@ -878,62 +874,81 @@ def load_model_hook(models, input_dir):
-    def compute_embeddings(prompt, text_encoders, tokenizers):
+    def compute_time_ids():
+        # Adapted from pipeline.StableDiffusionXLPipeline._get_add_time_ids
         original_size = (args.resolution, args.resolution)
```
```suggestion
        original_size = (args.resolution, args.resolution)
```
This should ideally be the original size of the passed image (before resizing), but ok to leave as is for now

patrickvonplaten commented on examples/dreambooth/README_sdxl.md:
```diff
@@ -164,6 +164,16 @@ Here's a side-by-side comparison of the with and without Refiner pipeline output
+
+### Specifying a better VAE
+
+SDXL's VAE is known to suffer from numerical instability issues. This is why we also expose a CLI argument namely `--pretrained_vae_model_name_or_path` that lets you specify the location of a better VAE (such as [this one](https://huggingface.co/madebyollin/sdxl-vae-fp16-fix)).
```
```suggestion
SDXL's VAE is known to suffer from numerical instability issues. This is why we also expose a CLI argument namely `--pretrained_vae_model_name_or_path` that lets you specify the location of a better VAE (such as [this one](https://huggingface.co/madebyollin/sdxl-vae-fp16-fix)).

```

Cool!

patrickvonplaten reviewed (approved):
Nice!

patrickvonplaten commented:
@williamberman ok for you?

williamberman commented on src/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl.py:
```diff
@@ -809,3 +810,66 @@ def __call__(
+            )
+
+    @classmethod
+    def save_lora_weights(
+        self,
+        save_directory: Union[str, os.PathLike],
+        unet_lora_layers: Dict[str, Union[torch.nn.Module, torch.Tensor]] = None,
+        text_encoder_lora_layers: Dict[str, Union[torch.nn.Module, torch.Tensor]] = None,
+        text_encoder_2_lora_layers: Dict[str, Union[torch.nn.Module, torch.Tensor]] = None,
+        is_main_process: bool = True,
+        weight_name: str = None,
+        save_function: Callable = None,
+        safe_serialization: bool = False,
+    ):
+        state_dict = {}
+
+        def pack_weights(layers, prefix):
+            layers_weights = layers.state_dict() if isinstance(layers, torch.nn.Module) else layers
+            layers_state_dict = {f"{prefix}.{module_name}": param for module_name, param in layers_weights.items()}
+            return layers_state_dict
+
+        state_dict.update(pack_weights(unet_lora_layers, "unet"))
+
+        if text_encoder_lora_layers and text_encoder_2_lora_layers:
+            state_dict.update(pack_weights(text_encoder_lora_layers, "text_encoder"))
+            state_dict.update(pack_weights(text_encoder_2_lora_layers, "text_encoder_2"))
+
```
<3

williamberman commented on examples/dreambooth/train_dreambooth_lora_sdxl.py:
```diff
@@ -420,46 +424,32 @@ def parse_args(input_args=None):
 
 
 class DreamBoothDataset(Dataset):
     """
     A dataset to prepare the instance and class images with the prompts for fine-tuning the model.
-    It pre-processes the images and the tokenizes prompts.
+    It pre-processes the images.
     """
 
     def __init__(
```
c'est magnifique

williamberman reviewed (approved):
perfect, lgtm!

sayakpaul commented:
Thanks all for your suggestions.
